### DISCUSSION TITLE
Code Base Review FY24

### DISCUSSION THREAD

**Discussion by psakievich** [Exawind/openturbine#108] [Q&A]
Discussion to organize thoughts and comments while reviewing the code base. I'm logging things as I see them here so I can have a reference. I will clean up my thoughts and comments as I go.

Process:

1. Docs
2. Tests
3. Src

I want to see how well I understand the ideas from the tests first before looking at the actual implementations given the goal for TDD in the project plan.

## **@psakievich - Overview:**
I see several positive things and potential for improvement. My high level summary is that the code's current form reads like a toolkit with a lot of fundamental math algorithms, and a few demonstration problems via unit-tests showing how you can piece the fundamental infrastructure together to solve a real problem.  

It all seems like a good start. It is clear a lot of care and thought has gone into designing the low level data structures and algorithms.

I would like to see the code begin to take a general form through the main function for problem setup and things like BC/IC/Forcing terms to be applied generally. This would help to understand how the problems will be setup and used.
It will also help when considering what a coupling interface would look like for talking with other codes (openfast, nalu-wind, amr-wind etc).

### Things I appreciate/like include:
- Variables are well named and it makes the code easier to read
- The low level math libraries are well tested and their use makes the algorithms easier to comprehend
- CI and CMake infrastructure are all looking good.
- Verification cases are built into the test suite and appear to be driving the development process
- Papers and references are in the code comments which make it easier to review and understand specific vernacular

### Things I would like to see improved:
- There is a lot of allocations between methods that obscures the flow of the code (see my comment on void  functions)
- Many of the tests are identical in function. Parameterization would help to comprehend the the extents of the testing better
- There is a lot of setup in the physics problem unit tests that would be nice to abstract into functions and methods.
- I see lots of hard coded values in the test suite. I would prefer to see them as variables unless there is a reference driving the value (i.e. from a paper). This will help with reduce duplicate code, as well as make the flow of the setup process and overall design.
- Documentation that shows the code working on device, convergence, verification results etc.  

### Things to anticipate and think about going forward
- Input deck file format and parser
- How abstract problems will be setup 
  - For example your current physics problems in the test suite would be nice to refactor or reproduce with a minimum input deck that you either hard code in a header or add as a first round of regression tests.
- API's for the abstract problems via input files and code to code communication (coupling with Nalu-Wind for example)
- Restarts, and interactions with a coupled aero solver via non-linear iterations
- Do we want to support variable time stepping? Can things be dynamically changed upon restart?

**Comment by psakievich** [comment]
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/docs/compiling.md?plain=1#L13-L15

Probably worth getting started with spack on this project since the rest of the exawind suite is based around using spack.

**Comment by psakievich** [comment]
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/tests/unit_tests/gen_alpha_poc/test_quaternions.cpp#L219-L234

Pretty wide range of tolerances here.  I am curious as to why, and if there are separate tolerances expected for single/double precision mode.  I would suggest making all similar tolerances a variable, and using something like `constexpr if` to set it appropriately based on precision/architecture.

I suspect that really understanding and monitoring the tolerances for these basic math operations will be important for everything else going forward.

**Reply by psakievich** [reply]
Similar question in this file. Are you building in the flexibility for multiple platforms with ranging tolerances etc?  Point of discussion, perhaps this precision is sufficient to satisfy the testing already.  
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/tests/unit_tests/gen_alpha_poc/test_heavy_top.cpp#L92-L102

**Reply by faisal-bhuiyan** [reply]
Recorded in https://github.com/Exawind/openturbine/issues/123.

**Comment by alanw0** [comment]
Low-level comment, initialization of views is sometimes a significant performance issue.
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/src/gen_alpha_poc/generalized_alpha_time_integrator.cpp#L143-L147
I think that Kokkos Views are zero-initialized by default, possibly meaning that the deep_copy with zero is un-necessary.
In fact a common performance optimization (sometimes significant) is to make sure Kokkos::View *doesn't* initialize if you know that you will be writing into it (not `+=`) and don't care what the initial value is, such as here:
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/src/gen_alpha_poc/generalized_alpha_time_integrator.cpp#L201-L202
In this case you would declare the Kokkos::View with the `Kokkos::WithoutInitializing` argument to avoid the internal kernel that is launched to perform the initialization.

**Reply by faisal-bhuiyan** [reply]
Thanks for the suggestion, @alanw0. Recorded as an issue in the repo: https://github.com/Exawind/openturbine/issues/118.

**Comment by alanw0** [comment]
Regarding passing 'raw' Kokkos::View types everywhere:
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/src/gen_alpha_poc/linearization_parameters.h#L14-L18
It's nice to see what the actual type is, but if performance optimizations or development leads to adding extra template parameters such as textured-memory (Kokkos::MemoryRandomAccess), or Atomic traits, then we may wish for central typedefs so that changes can be minimized. (These kinds of changes would manifest as extra template parameters in the Kokkos::View type.)

**Reply by deslaughter** [reply]
I think named types would be helpful since many of these have been refactored into multidimensional arrays where some of the dimensions are fixed size. It would remove the need to specify the dimension size in the type wherever it's being used.

**Comment by alanw0** [comment]
minor comment: should arguments be named in header declarations?
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/src/gen_alpha_poc/time_integrator.h#L21-L22
I don't know what the second argument is (size_t) unless I go find an override of this method in a derived class. It turns out (e.g. in generalized_alpha_time_integrator.cpp) that the size_t argument is `n_constraints`.

In general: the compiler doesn't require argument names for declarations, only types. I like argument names, although I recognize there is a "don't repeat yourself" viewpoint which says to only name the arguments in the cpp file where it is needed. I would suggest to just be consistent with the chosen style. There are 3 possible styles:
```
void foo(int);
void foo(int /*argname*/);
void foo(int argname);
```
I've seen all 3 of these styles in the Open Turbine code base.

**Reply by faisal-bhuiyan** [reply]
> I don't know what the second argument is (size_t) unless I go find an override of this method in a derived class. It turns out (e.g. in generalized_alpha_time_integrator.cpp) that the size_t argument is n_constraints.

I agree with this observation - I find argument names to be helpful for primitive types.

>  I would suggest to just be consistent with the chosen style.

Looks like we need to nail down the style choice in OpenTurbine - will start a style guide discussion to handle this.

**Reply by faisal-bhuiyan** [reply]
Transferred as https://github.com/Exawind/openturbine/issues/122

**Comment by alanw0** [comment]
The large number of unit tests is great, this is a nice artifact of doing test driven development. As development proceeds towards having an API for the "entry point" where a calling application would use Open Turbine, I would recommend creating more "doc" tests in addition to unit tests. Unit tests often check many aspects of correctness and "corner cases", and may not be perfectly readable for all audiences.
Doc tests should not add additional test coverage but rather act as a demonstration or example of how a client should interact with the library.
This test seems close to what I might consider a doc test:
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/tests/unit_tests/gen_alpha_poc/test_generalized_alpha_solver.cpp#L58

**Reply by faisal-bhuiyan** [reply]
@alanw0  Copy the above, will try to keep the (additional) utility of unit tests to work as a documentation for the codebase in mind. Please feel free to share some thoughts on the details and an example or two (in a codebase), if possible.

**Comment by psakievich** [comment]
Minor nitpick: some of these tests would be good candidates for a parameterization refactor.
Example:
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/tests/unit_tests/gen_alpha_poc/test_math_utilities.cpp#L8-L30

**Reply by faisal-bhuiyan** [reply]
Totally agree with the above - will create an issue to take care of this.

**Reply by faisal-bhuiyan** [reply]
Recorded here in https://github.com/Exawind/openturbine/issues/115.

**Comment by psakievich** [comment]
GLL quadrature is used a lot in the tests. I would just make a class for it. Seems like that is what you will use in the actual end application once you set it up as well.
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/src/gebt_poc/quadrature.h#L23

**Reply by michaelasprague** [reply]
As a point of clarification, we will use other quadrature methods as well (most likely trapezoidal when we have variable properties).

**Reply by faisal-bhuiyan** [reply]
I agree - we should add the GLL quadrature as a derived class from `Quadrature` and use that in the test suite to avoid repetition.

**Reply by faisal-bhuiyan** [reply]
Follow up issue: https://github.com/Exawind/openturbine/issues/116

**Comment by psakievich** [comment]
https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/src/gebt_poc/solver.cpp#L172-L264

If you can get away with not using `void` functions to update parameters it really helps with readability.  This pattern is repeated all throughout the code. It is hard to know what is getting updated and what isn't in these types of functions.

In the instance above, I would prefer a return of the residual in this case so it is clear what the output is.  Also, in this case you are zero'ing out the residual anyways so it doesn't seem like you **need** to pass by reference. There is an argument for saving on memory allocations, but looking at the utilization of these and the memory is getting allocated right before calling it:

https://github.com/Exawind/openturbine/blob/45aab4dbf501976640405be9250204455447dcd6/src/gebt_poc/clamped_beam.cpp#L75-L78

Which I would prefer to see as:
```c++
    auto residual_gen_coords = ElementalStaticForcesResidual(
        position_vectors_, gen_coords_1D, stiffness_matrix_, quadrature_);
```

**Reply by faisal-bhuiyan** [reply]
I agree with the above, returning a type from a function is helpful as a reader to understand what's being done in the function. I can take care of it as part of the refactor.

@deslaughter Could you remind me of your point regarding the above? I remember we discussed the above issue in our 2/20 meeting but cannot recall what it was.

**Reply by faisal-bhuiyan** [reply]
Referenced in the following issue: https://github.com/Exawind/openturbine/issues/117.

**Comment by psakievich** [comment]
Group Discussion: @rafmudaf @deslaughter @psakievich @alanw0 

Languages/Core Libraries (C++/Kokkos/Julia):
- Overall none of the reviewers see a compelling reason to go to Julia.  The cost of adding in an entirely new ecosystems of build dependencies seems like a high cost
- Kokkos Kernels were difficult for debugging and lacked a lot of linear algebra features @deslaughter 
  - Prototyped in rust and that seems viable.  There are GPU libraries available, but we do not have active expertise there.
- Rather than looking for a new technology perhaps we should focus on using the libraries and technology we already have expertise in well/better @rafmudaf. Specifically looking to Kokkos+Hypre(for sparse linear systems)
- Linear algebra:  seems like what we need is a clean API to translate the matrices to make it easier to port in and test various solver libraries.

**Reply by deslaughter** [reply]
I would like to keep the Rust development gonig as I've found it immensely helping for debugging the theory and prototyping the code. I also think it could serve as a good comparison for performance, in preliminary testing it's already slightly faster than the BeamDyn implementation in OpenFAST. As mentioned above, there are linear algebra libraries available such as [ArrayFire](https://arrayfire.com/) which includes GPU acceleration and sparse matrix operations (also has C++ and Python APIs).

**Reply by psakievich** [reply]
This seems reasonable to me. The main concern I would have is the added infrastructure and context switching a new language and libraries provide. So I would just say keep that in mind. If the rust implementation proves to be significantly easier then I would ask to work out how to deploy it via spack on multiple machines before committing to a switch down the road.

**Comment by alanw0** [comment]
Question from Mike: Thoughts on the overall structure/organization of the code.

**Reply by psakievich** [reply]
The organization seems mostly logical, but its current organization seems more like a tool kit than an application.  Specifically, the unit-tests show how concrete problems can be solved by assembling the tool kit components, but the abstraction of a complete problem (i.e read data, create objects, apply bc's forces etc, solve) is missing.

I would like to see the code reach a level of maturity where an arbitrary problem can be specified with a minimal set of information. Something like:

``` yaml
problem:
 name: "clamped beam"
 beam_positions:
 - start_position: 0, 0, 0
   end_position: 1, 0, 0
 boundary_conditions:
 - blah
 time_integration:
 - start_time: 0
 - end_time: 1
 - time_step: 1e-3
 solver_settings:
   type: kokkos_kernels # could be expanded to something else
   tolerances:
      absolute: 1e-12
      relative: 1e-7
   preconditioner_params:
   - doe
```

The details above are not important, or rather not meant to be a precise description, more that the idea of what is the minimal working set of data to describe the problem has been explored and captured in a clean form.  This will allow for things like solver abstractions (adding hypre, trilinos, petsc etc), different quadrature types.

Once a first draft of this is in place I think we will have a lot more to discuss regarding structure.

I would encourage the team to write out examples of what they would want to put into an input file to describe these problems and then explore what it would take to turn that into code.
Anything that can be a parameter should be, but also utilize defaults so the final form of the input deck (and eventual API) is not overwhelming and pedantic. 

Realistically parameterizing everything is a lot of work though. If you design your parsing and data structures (use structs etc) to allow for backward compatible additions then you don't need to define all the parameters at once. You can mark them as parametrizable in the code with a comment or something and turn them into a parameter with the default being set to what the hard coded option was to preserve backwards compatibility. Omission of new parameters in the input deck and/or API preserves old behavior.

**Reply by deslaughter** [reply]
A significant refactoring of the code is needed to minimize the dynamic allocation of Kokkos Views. Since the size of the problem doesn't change after initialization, all of the system matrices could be allocated once and subviews passed to relevant functions. The GEBT interpolation (shape function) arrays can be calculated at initialization and stored in matrices which can then interpolate the coordinates with matrix multiplication instead of individual operations. The Rust [prototype](https://github.com/deslaughter/roturb) was structured to minimize allocations and may be a good reference for the matrix sizing and operations. I think some performance profiling would be helpful to see which parts are being called repeatedly that may not need to be. 

I agree with @psakievich's comment that the code seems more like a toolkit than a program. Do we want to have the ability to use it as a toolkit with the main program being geared towards modeling a turbine? This approach may help us keep the elements and solver more general and extensible. Even if we don't, one of the next steps will need to be defining the representation of the turbine model in OpenTurbine, which could then guide how the element and solver API is structured.

Personally, I'd like to clean up the existing code and generalize the element and solver APIs so that the turbine model can be constructed in a generic manor. It has been difficult to represent more complex turbine designs (multi-rotor) in OpenFAST because of its design assumptions.

**Reply by michaelasprague** [reply]
@deslaughter is the refactor to minimize dynamic allocation something you are interested in undertaking? 

Regarding the toolkit aspects, I like that the work so far is viewed as a toolkit -- we can now add the layer(s) necessary to bring the parts together for a HAWT simulation -- but I don't want hinder our ability to build other systems.

**Reply by deslaughter** [reply]
Yes, I'd like to take a stab at doing the refactorization. I'm not sure when I can fit it in, let's discuss next week.

**Comment by alanw0** [comment]
Question from Mike: Thoughts on the code's readiness for an API for coupling to other codes.

**Reply by psakievich** [reply]
I detailed a lot of thoughts on this [here](https://github.com/Exawind/openturbine/discussions/108#discussioncomment-8444618), but I will expand on the idea related to coupling.

The ability to express a minimal problem is critical for a successful API. The steps I see are:
1) Identify the minimum input data needed for openturbine to do it's work
2) Assemble data structures that can be included via headers to express this information in simple types
3) Make this headers into a header only library that both codes can include

Until the data is cleanly assembled for running arbitrary classes of problems on the openturbine side, it is probably too early to be discuss much more on the API.

**Comment by rafmudaf** [comment]
I considered OpenTurbine from a high level perspective. After our group discussion (summarized in https://github.com/Exawind/openturbine/discussions/108#discussioncomment-8410729), I got more insight into the current state of OpenTurbine.

In particular, the architecture of the software as a wind turbine analysis tool has not yet been developed. My primary feedback was similar to https://github.com/Exawind/openturbine/discussions/108#discussioncomment-8444618 in that there are currently the building blocks for a cohesive software but no clear architecture yet. It sounds like now is the time to begin addressing this, and I recommend to step back and consider the design of this from a top-down, holistic perspective.

I also noticed there _is_ a particular coding style (i.e. the trailing `_` on properties, the `GetXYZ` methods, the place where long lines wrap, etc), but I don't see this style documented. As the number of developers scale, this will become more important. I suggest to start a style guide discussion to capture these low-level decisions with the understanding that it will change over time. See https://github.com/NREL/floris/discussions/292 for an idea of how this could be structured.

Is doxygen configured? I see the [doxygen directory](https://github.com/Exawind/openturbine/tree/main/docs/doxygen), but I don't see an entry point in the documentation site.

Similar to https://github.com/Exawind/openturbine/discussions/108#discussioncomment-8399856, I suggest to add some reporting on the problems you've solved so far. I would treat this as verification documentation, so add a static page to the docs describing the problem, how OpenTurbine was extended to solve the problem, and show the solution compared to any analytical results. Include a mesh dependency and convergence analysis.

**Reply by faisal-bhuiyan** [reply]
> I also noticed there is a particular coding style (i.e. the trailing _ on properties, the GetXYZ methods, the place where long lines wrap, etc), but I don't see this style documented. As the number of developers scale, this will become more important. I suggest to start a style guide discussion to capture these low-level decisions with the understanding that it will change over time. See https://github.com/NREL/floris/discussions/292 for an idea of how this could be structured.

Thanks for the suggestion - I agree the style choices might appear random to the reader. Started a discussion on this here:
https://github.com/Exawind/openturbine/discussions/125
